File: library/js/Native.js
const length = a => a.length;

const textLength = text => text.length;

const slice = (text, start, end) => text.slice(start, end);

const isArray = v => typeof(v) === typeof([]);

const unsafeIndex = (i, a) => a[i];

const unsafeCoerce = x => x;

const unsafeInsert = (a, e) => {
    a.push(e);
    return a;
};

const unsafeSet = (a, i, v) => {
    a[i + 1] = v;
    return a;
}

const toString = x => x.toString();

const print = x => console.log(x);

const error = x => {
    throw x;
};

const toLowerCase = x => x.toLowerCase();

const sin = x => Math.sin(x);

const cos = x => Math.cos(x);

const tan = x => Math.tan(x);

const toNumber = x => parseFloat(x);

const sequenceCompare = (compare, left, right) => {
    // `isArray` will be true for objects and arrays.
    // This is fine for this use case where compare should work on both.
    if (!(isArray(left) && isArray(right))) {
        return false
    }

    for (const key in left) {
        if (!compare(left[key])(right[key])) {
            return false
        }
    }

    for (const key in right) {
        if (!(key in left)) {
            return false
        }
    }

    return true
}

export default {
    length,
    textLength,
    isArray,
    unsafeIndex,
    slice,
    unsafeCoerce,
    unsafeInsert,
    unsafeSet,
    toLowerCase,
    sin,
    cos,
    tan,
    toNumber,
    toString,
    print,
    error,
    sequenceCompare,
}

File: library/lua/Native.lua
local utf8 = require("utf8")

local Native = {}

Native.length = function(a)
    return #a
end

Native.isArray = function(t)
    return type(t) == "table"
end

Native.unsafeIndex = function(i, a)
    return a[i + 1]
end

Native.unsafeCoerce = function(x)
    return x
end

Native.unsafeInsert = function(a, e)
    table.insert(a, e)
    return a
end

Native.unsafeSet = function(a, i, v)
    a[i + 1] = v
    return a
end

Native.sin = math.sin
Native.cos = math.cos
Native.tan = math.tan
Native.print = print

Native.toString = tostring

Native.toLowerCase = function(t)
    string.lower(s)
end

Native.textLength = function(s)
    return utf8.len(s)
end

Native.slice = function(s, startIndex, endIndex)
    local startOffset = utf8.offset(s, startIndex + 1)
    local endOffset = utf8.offset(s, endIndex + 1)

    if startOffset == nil then
        return ""
    end

    if endOffset == nil then
        return string.sub(s, startOffset)
    end

    return string.sub(s, startOffset, endOffset - 1)
end

Native.toNumber = function(s)
    return tonumber(s) or error("\"" .. s .. "\" is not a number")
end

Native.error = error

return Native
